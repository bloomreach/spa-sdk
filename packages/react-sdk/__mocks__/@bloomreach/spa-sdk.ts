/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

const spaSdk = jest.requireActual('@bloomreach/spa-sdk');

export const META_POSITION_BEGIN = spaSdk.META_POSITION_BEGIN;
export const META_POSITION_END = spaSdk.META_POSITION_END;

export function mockMeta(data: string, position: string) {
  return {
    getData: jest.fn().mockReturnValue(data),
    getPosition: jest.fn().mockReturnValue(position),
  };
}

export function mockNoCommentMeta() {
  return mockMeta('not-a-comment', META_POSITION_BEGIN);
}

const componentMock = {
  getMeta: jest.fn().mockReturnValue([
    mockMeta('meta-begin', META_POSITION_BEGIN),
    mockMeta('meta-end', META_POSITION_END),
  ]),
};

export const pageMock = {
  getComponent: jest.fn().mockReturnValue(componentMock),
  getContent: jest.fn(),
  getTitle: jest.fn(),
  sync: jest.fn(),
};

export const initialize = jest.fn();
export const destroy = jest.fn();

export function isMetaComment(value: any) {
  return value.getData() !== 'not-a-comment';
}
