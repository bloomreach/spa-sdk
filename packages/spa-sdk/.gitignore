/coverage
/docs
/lib
/node_modules
.DS_store