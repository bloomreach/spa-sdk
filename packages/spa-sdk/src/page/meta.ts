/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const META_POSITION_BEGIN = 'begin';
export const META_POSITION_END = 'end';

export type MetaType = string;
export type MetaPosition = typeof META_POSITION_BEGIN | typeof META_POSITION_END;

export interface MetaModel<T extends MetaType = MetaType> {
  data: string;
  type: T;
}

export interface MetaCollectionModel {
  beginNodeSpan?: MetaModel[];
  endNodeSpan?: MetaModel[];
}

/**
 * Meta information describing a part of the page.
 */
export interface Meta<T extends MetaType = MetaType> {
  /**
   * Returns meta-data.
   */
  getData(): string;

  /**
   * Returns meta-data type.
   */
  getType(): T;

  /**
   * Returns meta-data position relative to the related element.
   */
  getPosition(): MetaPosition;
}

export class Meta<T extends MetaType = MetaType> implements Meta {
  constructor(protected model: MetaModel<T>, protected position: MetaPosition) {}

  getData() {
    return this.model.data;
  }

  getType() {
    return this.model.type;
  }

  getPosition() {
    return this.position;
  }
}
