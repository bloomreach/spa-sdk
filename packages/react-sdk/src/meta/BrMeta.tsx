/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import ReactDOM from 'react-dom';
import { MetaComment, Meta, META_POSITION_BEGIN, META_POSITION_END, isMetaComment } from '@bloomreach/spa-sdk';

interface BrMetaProps {
  meta: Meta[];
}

export class BrMeta extends React.Component<BrMetaProps> {
  private comments: Comment[] = [];
  private head?: Element | Text;
  private tailRef = React.createRef<HTMLElement>();

  componentDidMount() {
    this.head = ReactDOM.findDOMNode(this)!;
    this.renderMeta();
  }

  componentDidUpdate() {
    this.head = ReactDOM.findDOMNode(this)!;
    this.removeMeta();
    this.renderMeta();
  }

  componentWillUnmount() {
    this.removeMeta();
  }

  private removeMeta() {
    this.comments.splice(0).forEach(comment => comment.remove());
  }

  private renderMeta() {
    this.props.meta.filter(meta => isMetaComment(meta) && meta.getPosition() === META_POSITION_BEGIN)
      .forEach(this.renderMetaComment.bind(this));
    this.props.meta.filter(meta => isMetaComment(meta) && meta.getPosition() === META_POSITION_END)
      .reverse()
      .forEach(this.renderMetaComment.bind(this));
  }

  private renderMetaComment(meta: MetaComment) {
    const comment = this.head!.ownerDocument!.createComment(meta.getData());
    this.comments.push(comment);

    if (meta.getPosition() === META_POSITION_BEGIN) {
      return void this.head!.parentNode!.insertBefore(comment, this.head!);
    }

    if (!this.tailRef.current!.nextSibling) {
      return void this.tailRef.current!.parentNode!.appendChild(comment);
    }

    this.tailRef.current!.parentNode!.insertBefore(comment, this.tailRef.current!.nextSibling);
  }

  render() {
    return (
      <>
        {this.props.children}
        {this.props.meta.length ? <span style={{ display: 'none' }} ref={this.tailRef} /> : null}
      </>
    );
  }
}
