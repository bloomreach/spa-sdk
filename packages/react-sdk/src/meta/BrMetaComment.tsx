/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { MetaComment } from '@bloomreach/spa-sdk';

interface BrMetaCommentProps {
  meta: MetaComment;
}

export class BrMetaComment extends React.Component<BrMetaCommentProps> {
  private placeholder = React.createRef<HTMLElement>();
  private comment?: Comment;

  componentDidMount() {
    const placeholder = this.placeholder.current!;
    this.comment = placeholder.ownerDocument!.createComment(this.props.meta.getData());
    placeholder.parentNode!.replaceChild(this.comment, placeholder);
  }

  shouldComponentUpdate(nextProps: BrMetaCommentProps) {
    return nextProps.meta.getData() !== this.props.meta.getData();
  }

  componentDidUpdate() {
    this.comment!.replaceData(0, this.comment!.data.length, this.props.meta.getData());
  }

  componentWillUnmount() {
    this.comment!.parentNode!.replaceChild(this.placeholder.current!, this.comment!);
    delete this.comment;
  }

  render() {
    return <span style={{ display: 'none' }} ref={this.placeholder} />;
  }
}
