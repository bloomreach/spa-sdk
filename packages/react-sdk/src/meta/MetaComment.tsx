/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { MetaComment as SpaMetaComment } from '@bloomreach/spa-sdk';

interface MetaCommentProps {
  meta: SpaMetaComment;
}

export class MetaComment extends React.Component<MetaCommentProps> {
  private placeholder = React.createRef<HTMLElement>();
  private comment?: Comment;

  componentDidMount() {
    if (!this.placeholder.current) {
      return;
    }

    const placeholder = this.placeholder.current;
    this.comment = placeholder.ownerDocument!.createComment(this.props.meta.getData());
    placeholder.parentNode!.replaceChild(this.comment, placeholder);
  }

  shouldComponentUpdate(nextProps: MetaCommentProps) {
    return nextProps.meta.getData() !== this.props.meta.getData();
  }

  componentDidUpdate() {
    this.comment!.replaceData(0, this.comment!.data.length, this.props.meta.getData());
  }

  componentWillUnmount() {
    if (!this.comment) {
      return;
    }

    this.comment.remove();
  }

  render() {
    return <span style={{ display: 'none' }} ref={this.placeholder} />;
  }
}
