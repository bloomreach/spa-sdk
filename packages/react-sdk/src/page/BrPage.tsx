/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { destroy, initialize, Configuration, Page } from '@bloomreach/spa-sdk';
import { BrMappingContext, BrNode } from '../component';
import { BrPageContext } from './BrPageContext';

interface BrPageProps {
  configuration: Configuration;
  mapping: React.ContextType<typeof BrMappingContext>;
}

interface BrPageState {
  page?: Page;
}

export class BrPage extends React.Component<BrPageProps, BrPageState> {
  constructor(props: BrPageProps) {
    super(props);

    this.state = {};
  }

  componentDidMount() {
    this.initializePage();
  }

  componentDidUpdate(prevProps: BrPageProps, prevState: BrPageState) {
    if (this.props.configuration !== prevProps.configuration) {
      this.destroyPage();
      this.initializePage();
    }

    if (this.state.page && this.state.page !== prevState.page) {
      this.state.page.sync();
    }
  }

  componentWillUnmount() {
    this.destroyPage();
  }

  private async initializePage() {
    try {
      const page = await initialize(this.props.configuration);
      this.setState({ page });
    } catch (error) {
      this.setState(() => { throw error; });
    }
  }

  private destroyPage() {
    if (!this.state.page) {
      return;
    }

    destroy(this.state.page);
  }

  render () {
    if (!this.state.page) {
      return null;
    }

    return (
      <BrPageContext.Provider value={this.state.page}>
        <BrMappingContext.Provider value={this.props.mapping}>
          <BrNode component={this.state.page.getComponent()}>
            {this.props.children}
          </BrNode>
        </BrMappingContext.Provider>
      </BrPageContext.Provider>
    );
  }
}
