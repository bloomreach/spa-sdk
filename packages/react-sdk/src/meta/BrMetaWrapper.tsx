/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { Meta, META_POSITION_BEGIN, META_POSITION_END } from '@bloomreach/spa-sdk';
import { BrMeta } from './BrMeta';

interface BrMetaWrapperProps {
  meta: Meta[];
}

function renderMeta(meta: Meta[], position: typeof META_POSITION_BEGIN | typeof META_POSITION_END) {
  return meta
    .filter(meta => position === meta.getPosition())
    .map((meta, index) => <BrMeta key={index} meta={meta} />);
}

// tslint:disable-next-line:function-name
export function BrMetaWrapper(props: React.PropsWithChildren<BrMetaWrapperProps>) {
  return (
    <>
      {renderMeta(props.meta, META_POSITION_BEGIN)}
      {props.children}
      {renderMeta(props.meta, META_POSITION_END)}
    </>
  );
}
