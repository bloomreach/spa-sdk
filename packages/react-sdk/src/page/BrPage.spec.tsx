/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as spaSdk from '@bloomreach/spa-sdk';

import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';

import { BrPage } from './BrPage';
import { Meta } from '../meta/Meta';

const model = {
  data: {
    page: {
      id: 'page1',
      type: 'COMPONENT',
      name: 'Page1',
      "_meta": {
        "beginNodeSpan": [
          {
            "type": "comment",
            "data": "comment-start"
          },
        ],
        "endNodeSpan": [
          {
            "type": "comment",
            "data": "comment-end"
          },
        ],
      },
    },
  }
};

const config = {
  httpClient: jest.fn(async () => model),
  request: { path: '/' },
  options: {
    live: {
      pageModelBaseUrl: 'http://localhost:8080/site/my-spa/resourceapi',
    },
    preview: {
      pageModelBaseUrl: 'http://localhost:8080/site/_cmsinternal/my-spa/resourceapi',
      spaBasePath: '/site/_cmsinternal/my-spa',
    },
  },
};

class TestComponent extends React.Component {}
const mapping = { TestComponent };

describe('BrPage', function() {
  let spaInit: any;
  let spaDestroy: any;

  beforeEach(() => {
    spaInit = jest.spyOn(spaSdk, 'initialize');
    spaDestroy = jest.spyOn(spaSdk, 'destroy');
  });

  afterEach(() => {
    spaInit.mockRestore();
    spaDestroy.mockRestore();
  });

  async function createBrPage(children: JSX.Element | undefined = undefined): Promise<ShallowWrapper> {
    const wrapper = shallow(
      <BrPage configuration={config} mapping={mapping}>
        {children}
      </BrPage>
    );
    // wait for BrPage.initializePage to resolve
    const page = await spaInit.mock.results[0].value;

    return wrapper;
  }

  it('should initialize the SPA SDK', async () => {
    const wrapper = await createBrPage();
    expect(spaInit).toHaveBeenCalledWith(config);

    const page: spaSdk.Page = wrapper.state('page');
    expect(page).toBeDefined();
    expect(page.getComponent().getId()).toBe('page1');
  });

  it('should render nothing if there is no page', async () => {
    spaInit.mockReturnValue(null);
    const wrapper = await createBrPage();

    expect(wrapper.isEmptyRender()).toBe(true);
  });

  it('should render nothing if there is an error loading the page', (done) => {
    spaInit.mockRejectedValue('error-loading-page');
    const wrapper = shallow(<BrPage configuration={config} mapping={{}}/>);
    spaInit.mock.results[0].value.catch((e: Error) => {
      expect(wrapper.isEmptyRender()).toBe(true);
      expect(e).toEqual('error-loading-page');
      done();
    });
  });

  it('should render MappingContext.provider', async () => {
    const wrapper = await createBrPage();

    expect(wrapper.find("ContextProvider").first().prop('value')).toEqual(mapping);
  });

  it('should render PageContext.provider', async () => {
    const wrapper = await createBrPage();
    const page = wrapper.state('page');

    expect(wrapper.find("ContextProvider").last().prop('value')).toEqual(page);
  });

  it('should render children', async () => {
    const wrapper = await createBrPage(<div id="br-page-child"></div>);
    await spaInit.mock.results[0].value;

    expect(wrapper.contains([
      <div id="br-page-child"></div>
    ])).toBe(true);
  });

  it('should render meta data in comments', async () => {
    const wrapper = await createBrPage();
    const page: spaSdk.Page = wrapper.state('page');
    const pageMeta = page.getComponent().getMeta();

    expect(wrapper.contains(<Meta meta={pageMeta[0]} />)).toBe(true);
    expect(wrapper.contains(<Meta meta={pageMeta[1]} />)).toBe(true);
  });

  it('should destroy current page and init new page when configuration changes', async () => {
    const wrapper = await createBrPage();
    const page = wrapper.state('page');
    const newConfig = { ...config };
    wrapper.setProps({
      configuration: newConfig,
    });

    expect(spaDestroy).toHaveBeenCalledWith(page);
    expect(spaInit).toHaveBeenCalledWith(newConfig);
  });

  it('should destroy the page when unmounting', async () => {
    const wrapper = await createBrPage();
    const page = wrapper.state('page');

    wrapper.unmount();
    expect(spaDestroy).toHaveBeenCalledWith(page);
  });

  it('should not destroy a null page when unmounting', async () => {
    spaInit.mockReturnValue(null);
    const wrapper = await createBrPage();

    wrapper.unmount();
    expect(spaDestroy).not.toHaveBeenCalled();
  });
});
