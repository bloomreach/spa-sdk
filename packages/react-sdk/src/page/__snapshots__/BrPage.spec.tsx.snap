// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrPage render should not render children if there is no page and NBR mode is false 1`] = `<DocumentFragment />`;

exports[`BrPage render should render children 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`BrPage render should render children if there is no page and NBR mode is true 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`BrPage render should render nothing if there is an error loading the page 1`] = `
<DocumentFragment>
  <div>
    fallback
  </div>
</DocumentFragment>
`;
